### PR TITLE
rgw: fix SLO/DLO range requests

### PR DESCRIPTION
**Pull request by shashalu**
Sending any SLO/DLO range requests will retrun InvalidRange error. SLO/DLO should call range_to_ofs after got object size instead of read_op.prepare().

Signed-off-by: Shasha Lu <lu.shasha@eisoo.com>

### CODE REVIEW

**Comment by shashalu** [comment]
@cbodley  hi, could you review this for me? Thanks

**Comment by cbodley** [comment]
@shashalu would you be willing to add some functional tests for this to https://github.com/ceph/swift/?

**Comment by shashalu** [comment]
@cbodley  there is already a function testRangedGets.  But I'm not clear how to run these tests.

**Comment by cbodley** [comment]
@shashalu testRangedGets doesn't test against slo/dlo objects

here's how i run the functional tests using a vstart cluster:
> ~/ceph/build $ OSD=1 MON=1 RGW=1 MDS=0 MGR=0 ../src/vstart.sh -n -d

modify the sample config file at ~/swift/test/functional/sample.conf to point at radosgw port 8000:
> sed 's/8080/8000/' sample.conf > rgw.conf

run the subset of tests that pass against rgw:
> ~/swift/test/functional $ SWIFT_TEST_CONFIG_FILE=rgw.conf nosetests -v -a '!fails_on_rgw'

**Comment by shashalu** [comment]
@cbodley  Thanks, i'll try that.

**Comment by shashalu** [comment]
@cbodley  add functional tests for this ceph/swift#3

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -1250,8 +1250,21 @@ int RGWGetObj::handle_user_manifest(const char *prefix)
...
 
+  r = RGWRados::Object::Read::range_to_ofs(s->obj_size, ofs, end);
+  if (r < 0) {
+    return r;
+  }
+
+  r = iterate_user_manifest_parts(s->cct, store, ofs, end,
+        pbucket_info, obj_prefix, bucket_acl, *bucket_policy,
+        &total_len, nullptr, nullptr,
+        nullptr, nullptr);
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: this extra call to `iterate_user_manifest_parts()` is needed to support negative offsets?

Comment (shashalu) [inline src/rgw/rgw_op.cc]: yes.  For negative offsets we should get the slo/dlo obj_size first.  Then fix the range.
The first `iterate_user_manifest_parts()` get object's objsize. Then fix the negative ranges.
The second `iterate_user_manifest_parts()` get total_len
The third `iterate_user_manifest_parts()` read data.

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -1250,8 +1250,21 @@ int RGWGetObj::handle_user_manifest(const char *prefix)
...
+  if (r < 0) {
+    return r;
+  }
+
```
Comment (shashalu) [inline src/rgw/rgw_op.cc]: yes, i'll revise this

**Pull request merged**